// configs/settings/index.js
/***
 * Copyright (c) 2015, 2016 Alex Grant (@localnerve), LocalNerve LLC
 * Copyrights licensed under the BSD License. See the accompanying LICENSE file for terms.
 *
 * Environment variables used:
 *   NODE_ENV
 *   ASSET_HOST
 *   APP_HOSTNAME
 *
 * Environmentally aware project dir, file, and settings.
 * Avoids non-module dir strings all over project.
 * Dir and file paths are completed to a given baseDir, so can change relative
 * point of reference.
 *
 * Creates a settings configuration with three major areas:
 *   config.src - The source dirs and files.
 *   config.output - The compilation target source dirs and files.
 *   config.dist - The build output destination dirs and files. These become
 *     server-side public assets (js, html, css, images, json, xml, etc...).
 *   config.web - files, dirs, and settings from the web client perspective.
 */
'use strict';

var path = require('path');
var merge = require('lodash/merge');
var assets = require('./assets');
var utils = require('./utils');

var distbase = 'dist';
var srcbase = 'src';
var assetsDir = 'assets';
var applicationDir = 'application';
var publicbase = '/public';

var prependPathToObject = utils.prependPathToObject;

/***
 * Directories and files that are in src, distribution, and web
 */
var commonDirs = {
  images: 'images',
  styles: 'styles',
  fonts: 'fonts',
  scripts: 'scripts'
};
var commonFiles = {
  five00: '500.html',
  five03: '503.html',
  favicon: path.join(commonDirs.images, 'favicon.ico'),
  robotsTemplate: 'robots.txt',
  inlineScript: path.join(commonDirs.scripts, 'inline.js'),
  appManifest: 'manifest.json',
  browserConfig: 'browserconfig.xml'
};

/***
 * Directories and files that are in both dist and web
 */
var outputDirs = {
};
var outputFiles = {
  css: {
    inline: path.join(commonDirs.styles, 'index.css'),
    other: [
      // add other external stylesheets to async load at startup
      // path.join(commonDirs.styles, 'settings.css')
    ]
  }
};

/***
 * Source only dirs and files
 */
var srcDirs = {
  application: applicationDir,
  components: path.join(applicationDir, 'components'),
  assets: path.join(applicationDir, assetsDir),
  configs: 'node_modules/configs',
  scripts: path.join(applicationDir, assetsDir, commonDirs.scripts),
  tests: 'tests'
};
var srcFiles = {
  assetsJson: path.join(
    srcDirs.configs, path.basename(__dirname), assets.assetsJsonFile
  ),
  clientEntry: path.join(srcDirs.application, 'client.js'),
  serviceWorker: {
    registration: path.join(srcDirs.scripts, 'service-worker-registration.js'),
    precache: path.join(srcDirs.scripts, 'sw', 'precache.js'),
    data: path.join(srcDirs.scripts, 'sw', 'data.json'),
    entry: path.join(srcDirs.scripts, 'sw', 'index.js')
  }
};

/**
 * Get settings to override by environment.
 * production override uses ASSET_HOST env variable.
 *
 * @param {String} env - The environment string.
 * @param {String} baseDir - The base directory.
 * @returns {Object|Undefined} The environment specific overrides or undefined if none.
 */
function overrides (env, baseDir) {
  var envOverrides = {
    production: {
      dist: {
        baseDir: path.join(baseDir, distbase, 'release')
      },
      loggerFormat: 'tiny',
      web: {
        // ssl: true, // for ssl here on this app (this tier)
        sslRemote: true, // for ssl elsewhere (an appliance or CloudFlare, etc)
        assetAge: 31556926000
      }
    }
  };
  return envOverrides[env];
}

/**
 * Make the settings configuration object.
 *
 * @param {Object} nconf - The nconfig object.
 * @returns The settings configuration object.
 */
function makeConfig (nconf) {
  var env = nconf.get('NODE_ENV');

  // Update relative baseDir if defined
  var baseDir = nconf.get('baseDir') || '.';

  /**
   * The exported settings config
   */
  var config = {
    dist: {
      baseDir: path.join(baseDir, distbase, 'debug')
    },
    src: {
      baseDir: path.join(baseDir, srcbase)
    },
    web: {
      baseDir: publicbase,
      assetAge: 0,
      assetHost:
        nconf.get('ASSET_HOST') || nconf.get('APP_HOSTNAME') || 'localhost',
      ssl: false,
      sslRemote: false,
      robots: '/robots.txt',
      sitemap: '/sitemap.xml',
      appHostname: nconf.get('APP_HOSTNAME') || 'localhost'
    },
    output: {
      baseDir: path.join(baseDir, 'output')
    },

    distbase: path.join(baseDir, distbase),
    vendor: {
      css: 'node_modules/foundation-apps/scss'
    },

    loggerFormat: 'dev'
  };

  //
  // Environment overrides
  merge(config, overrides(env, baseDir));

  // Assemble config.output
  Object.assign(
    config.output,
    prependPathToObject(srcDirs, config.output.baseDir),
    prependPathToObject(srcFiles, config.output.baseDir)
  );

  // Assemble config.src
  Object.assign(
    config.src,
    prependPathToObject(commonDirs, path.join(
      config.src.baseDir, srcDirs.assets
    )),
    prependPathToObject(commonFiles, path.join(
      config.src.baseDir, srcDirs.assets
    )),
    prependPathToObject(srcDirs, config.src.baseDir),
    prependPathToObject(srcFiles, config.src.baseDir)
  );

  // Assemble config.dist and config.web
  [config.dist, config.web].forEach(function (config) {
    Object.assign(
      config,
      prependPathToObject(commonDirs, config.baseDir),
      prependPathToObject(commonFiles, config.baseDir),
      prependPathToObject(outputDirs, config.baseDir),
      prependPathToObject(outputFiles, config.baseDir)
    );
  });

  // A little extra web-only config for determining built assets
  config.web.assets = assets.assetsConfig(config.web.scripts);

  return config;
}

module.exports = makeConfig;
